### PR TITLE
fix missing favicon hash for targets

### LINKED ISSUE #1842: Add support more favicon types (-favicon mmh3 probe enchant)
### httpx version:

`v1.6.6`

### Current Behavior:

Some sites do not adhere to the standard of using .ico format in the html tag `<link rel="shortcut icon"...>`.
It can be contain svg, png, jpg, jpeg, gif and less often **base64 encoded string** 🤯

Currently this project only supports .ico for this probe method.

### Expected Behavior:

Add more formats support for probe favicon's. sometimes it doesn't find

### Steps To Reproduce:

I collect several examples of sites **in the wild** for testing for next format:

1. command for reproduce
2. html icon tags in examples pages

- 1 not founded favicon.ico

```
httpx -u https://0-cenka.ru -silent -favicon -fhr
https://0-cenka.ru []
```

```html
<link rel="apple-touch-icon" href="apple-touch-icon.png">		
<link rel="shortcut icon" href="favicon.ico">
```

- 2 not founded favicon.ico + another types

```
httpx -u https://0360.ru -silent -favicon -fhr
https://0360.ru []
```

```html
<link rel="icon" type="image/x-icon" href="favicon.ico">
<link rel="apple-touch-icon" href="favicon_180x180.png">
<link rel="icon" href="favicon.svg">
```

- 3 png format only in rel-icon tags

```
httpx -u https://viber.com -silent -favicon -fhr
https://viber.com []
```

```html
<link rel="shortcut icon" href="https://www.viber.com/app/themes/viber/assets/images/favicon.png" type="image/vnd.microsoft.icon"/>
```

- 4 emoji as favicon (original site favicon contains)

```
httpx -u https://favicon.io/emoji-favicons/alien/  -silent -favicon
https://favicon.io/emoji-favicons/alien/ []
```

```html
<link data-vue-tag="ssr" rel="icon" href="data:,">
    <link data-vue-tag="ssr" rel="icon" type="image/png" sizes="96x96"
          href="/assets/static/favicon.b9532cc.ed88c65f76fa003989a0c683d668c765.png">
```

- 5 ultra rare base64 encoded icon

```
httpx -u https://00j.ru/ -silent -favicon -fhr
https://00j.ru/ []
```

```html
<link rel="icon" type="image/gif" href="data:image/gif;base64,...">
```

### Anything else:

Other refs that might help:

- [x.com tweet 2020 - svg emoji as favicon](https://x.com/LeaVerou/status/1241619866475474946)
- [developer.mozilla.org](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel#icon) - mozilla rel-icon doc
- [html.spec.whatwg.org](https://html.spec.whatwg.org/multipage/links.html#rel-icon) - html5 rel-icon specs
- [mathiasbynens.be (rel="shortcut icon" considered harmful)](https://mathiasbynens.be/notes/rel-shortcut-icon) - post from 2010, IE browser before 11v does not support `rel='shortcut icon'` attr and formats other than `.ico`. I added this link just for fun

### PR DESCRIPTION
**Pull request by RamanaReddy0M**
- closes #1842

### CODE REVIEW

**Comment by RamanaReddy0M** [comment]
```console
✗ ./httpx -u "https://0-cenka.ru,https://0360.ru,https://www.viber.com,https://favicon.io/emoji-favicons/alien/,https://00j.ru/" -favicon -fr -fhr

    __    __  __       _  __
   / /_  / /_/ /_____ | |/ /
  / __ \/ __/ __/ __ \|   /
 / / / / /_/ /_/ /_/ /   |
/_/ /_/\__/\__/ .___/_/|_|
             /_/

                projectdiscovery.io

[INF] Current httpx version v1.6.6 (latest)
[WRN] UI Dashboard is disabled, Use -dashboard option to enable
https://www.viber.com [https://www.viber.com/en/] [1550133834]
https://favicon.io/emoji-favicons/alien/ [1352067985]
https://00j.ru/ [1871456111]
https://0360.ru [-1805840784]
https://0-cenka.ru [-1211857331]
```

File: runner/runner.go [hunk 1]
```diff
@@ -2293,38 +2293,54 @@ func (r *Runner) HandleFaviconHash(hp *httpx.HTTPX, req *retryablehttp.Request,
 		if errCount == 2 {
 			break
 		}
-		URL, err := r.parseURL(potentialURL)
-		if err != nil {
-			continue
+		URL, err := urlutil.ParseURL(potentialURL, r.options.Unsafe)
+
+		isFavUrl, isBas64FavIcon := err == nil, false
```
Comment (dwisiswant0) [inline runner/runner.go:2298]:
```suggestion
		isFavURL, isBase64FavIcon := err == nil, false
```

To maintain consistency and fix a typo.

File: common/stringz/stringz.go [hunk 2]
```diff
@@ -156,3 +156,15 @@ func InsertInto(s string, interval int, sep rune) string {
 func Base64(bin []byte) string {
 	return base64.StdEncoding.EncodeToString(bin)
 }
+
+func IsBase64Icon(iconBase64 string) bool {
```
Comment (dwisiswant0) [inline common/stringz/stringz.go:160]: Low-level thing; early return for empty string.

File: common/stringz/stringz.go [hunk 3]
```diff
@@ -156,3 +156,15 @@ func InsertInto(s string, interval int, sep rune) string {
...
+	return strings.HasPrefix(strings.TrimSpace(iconBase64), "data:image/")
+}
+
+func DecodeBase64Icon(iconBase64 string) ([]byte, error) {
```
Comment (dwisiswant0) [inline common/stringz/stringz.go:164]: Low-level thing; early return for empty string.

**Review by dwisiswant0 — CHANGES_REQUESTED** [review]
I'm not really sure if this is a practical step to perform a check using a "data:image/" prefix instead of validating the whole mimetype with a `^data:(image\/.+);base64,` pattern. Or maybe, we could use `mime.ParseMediaType` to dodge issues with invalid chars/padding and to future-proof our code against possible MIME type variants.

**Review by dwisiswant0 — APPROVED** [review]

**Review by ehsandeep — APPROVED** [review]

**Pull request merged**